idempotent_download.go

```diff
@@ -1,18 +1,19 @@
 package main
 
 import (
+	"context"
 	"crypto/md5"
 	"encoding/hex"
+	"errors"
 	"github.com/sirupsen/logrus"
 	"io"
 	"os"
 )
 
 // Interface with logic to govern how to actually pull objects
 type downloader interface {
-	Download(remotePath, outputPath string) error
-	DownloadAndValidateChecksum(remotePath, outputPath, checksum string) error
```

Comment:
I've removed `DownloadAndValidateChecksum` because this logic is the same for both client types